kompile: add check for token sort productions

SchmErik opened the pull request:
Token sorts should only contain productions that have `token` or
`function` labels. Any other label (including subsorts) are illegal.

Fixes: #2119

Commit:
kompile: add check for token sort productions

Token sorts should only contain productions that have `token` or
`function` labels. Any other label (including subsorts) are illegal.

Commit:
invalidConstructor check: update test results

Commit:
regression-new: add tests for token check

radumereuta commented on kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java:
```diff
@@ -0,0 +1,43 @@
+public class CheckTokens {
+    private final Set<KEMException> errors;
+    private final Module m;
+    private static final ImmutableSet<String> ignoredSortNames = ImmutableSet.of("Bag", "IOFile", "IOInt", "IOString", "K", "KBott", "KItem", "KLabel", "KList", "Sort");
```
Why `Sort`? Where is that coming from?
Not sure about some others also. `KItem` for example.
I just noticed that in the original issue, I complained about the LLVM and Haskell backend. But Java is a bit different. Probably we don't need to run this check there.
Could you check all these, please?

SchmErik replied:
I came up with these by running kompile on the tokenCheck.k file and looking at the errors that were generated without ignoring special sort names and `Sort` was one of them

radumereuta replied:
You should check all of them. They might indicate something else is wrong.

SchmErik replied:
oh strange... Now I'm getting that I only need to ignore "KBott" and "KLabel"

SchmErik replied:
I think this was written when I was still working on solidifying the logic. I removed all but KBott and KLabel and I'm rerunning the testsuite locally. I'll push another commit addressing this and the other feedback you have. Thanks!

Commit:
fix typo

radumereuta commented on kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    // This check ensures that sorts containing token declarations only contain syntax declarations that are also tokens (or macros).
+    public void check(Sentence sentence) {
```
`private void check(Production p) {`
You can match directly on Production and do away with this check.

SchmErik replied:
ah, thanks that would make things more clear

radumereuta commented on kernel/src/main/java/org/kframework/compile/checks/CheckTokens.java:
```diff
@@ -0,0 +1,43 @@
+            return;
+        }
+
+        errors.add(KEMException.compilerError("Sort " + p.sort().name() + " was declared as a token. Productions of this sort can only contain [function] or [token] labels.", p));
```
Line too long.

Commit:
Address code review feedback

Commit:
pl-tutorial: Add token label to `Id` token sorts

These files contained token sorts called `Id`. The java backend seems
to allow token sorts to be defined with the `token` label or as strings.
The haskell and llvm backends only allow token sorts to have productions
that are tagged with the `token` label. Since the java backend will be
deprecated, update the sort productions to include the `token` label
so that it will work on the newer backends as well as the java backend.

Commit:
Merge branch 'master' into check-tokens-2119

Commit:
pl-tutorial: Add token label to `Name` token sorts

This is the same change that was done in the previous commit to the `Id` sort.

Commit:
kompile: do not run token checks when running java backend

The java backend is more forgiving when it comes to token sort
declarations. Remove this check when kompile is running with the java
backend.

Commit:
checkToken: fix logic

Commit:
Revert "pl-tutorial: Add token label to `Name` token sorts"

This reverts commit e6866c59848e1256a3b4cf4355a38d12199863a7.

Commit:
Revert "pl-tutorial: Add token label to `Id` token sorts"

This reverts commit 7493722f473a987c9ef178368d84668a57cc2c91.

Commit:
pl-tutorial: add token labels to files kompiled using llvm backend

radumereuta committed:
Merge branch 'master' into check-tokens-2119

SchmErik commented:
this PR got a little hectic with a bunch of commits that ended up getting reverted or were typo fixes. Now that the CI is passing, I'll open yet another PR with cleaner commits.
